File: core/src/pl/shockah/godwit/geom/Triangle.java
package pl.shockah.godwit.geom;

import com.badlogic.gdx.graphics.glutils.ShapeRenderer;

import javax.annotation.Nonnull;

import lombok.EqualsAndHashCode;
import pl.shockah.godwit.geom.polygon.Polygon;
import pl.shockah.godwit.geom.polygon.Polygonable;
import pl.shockah.godwit.gl.Gfx;
import pl.shockah.unicorn.Math2;
import pl.shockah.unicorn.ease.Easable;

@EqualsAndHashCode(callSuper = false)
public class Triangle extends AbstractShape implements Polygonable, Shape.Filled, Shape.Outline, Easable<Triangle> {
	@Nonnull public MutableVec2 point1;
	@Nonnull public MutableVec2 point2;
	@Nonnull public MutableVec2 point3;

	public Triangle(float x1, float y1, float x2, float y2, float x3, float y3) {
		this(new MutableVec2(x1, y1), new MutableVec2(x2, y2), new MutableVec2(x3, y3));
	}

	public Triangle(float x1, float y1, @Nonnull IVec2 point2, float x3, float y3) {
		this(new MutableVec2(x1, y1), point2, new MutableVec2(x3, y3));
	}

	public Triangle(@Nonnull IVec2 point1, float x2, float y2, float x3, float y3) {
		this(point1, new MutableVec2(x2, y2), new MutableVec2(x3, y3));
	}

	public Triangle(float x1, float y1, float x2, float y2, @Nonnull IVec2 point3) {
		this(new MutableVec2(x1, y1), new MutableVec2(x2, y2), point3);
	}

	public Triangle(float x1, float y1, @Nonnull IVec2 point2, @Nonnull IVec2 point3) {
		this(new MutableVec2(x1, y1), point2, point3);
	}

	public Triangle(@Nonnull IVec2 point1, float x2, float y2, @Nonnull IVec2 point3) {
		this(point1, new MutableVec2(x2, y2), point3);
	}

	public Triangle(@Nonnull IVec2 point1, @Nonnull IVec2 point2, float x3, float y3) {
		this(point1, point2, new MutableVec2(x3, y3));
	}

	public Triangle(@Nonnull IVec2 point1, @Nonnull IVec2 point2, @Nonnull IVec2 point3) {
		this.point1 = point1.mutableCopy();
		this.point2 = point2.mutableCopy();
		this.point3 = point3.mutableCopy();
	}

	@Override
	@Nonnull public Triangle copy() {
		return new Triangle(point1, point2, point3);
	}

	@Override
	public String toString() {
		return String.format("[Triangle: %s-%s-%s]", point1, point2, point3);
	}

	@Override
	@Nonnull public Rectangle getBoundingBox() {
		float minX = Math2.min(point1.x, point2.x, point3.x);
		float minY = Math2.min(point1.y, point2.y, point3.y);
		float maxX = Math2.max(point1.x, point2.x, point3.x);
		float maxY = Math2.max(point1.y, point2.y, point3.y);
		return new Rectangle(minX, minY, maxX - minX, maxY - minY);
	}

	@Override
	public void translate(@Nonnull IVec2 v) {
		point1.x += v.x();
		point1.y += v.y();
		point2.x += v.x();
		point2.y += v.y();
		point3.x += v.x();
		point3.y += v.y();
	}

	@Override
	public void mirror(boolean horizontally, boolean vertically) {
		if (horizontally) {
			point1.x *= -1;
			point2.x *= -1;
			point3.x *= -1;
		}
		if (vertically) {
			point1.y *= -1;
			point2.y *= -1;
			point3.y *= -1;
		}
	}

	@Override
	public void scale(float scale) {
		point1.set(point1.multiply(scale));
		point2.set(point2.multiply(scale));
		point3.set(point3.multiply(scale));
	}

	private static float sign(float x1, float y1, float x2, float y2, float x3, float y3) {
		return (x1 - x3) * (y2 - y3) - (x2 - x3) * (y1 - y3);
	}

	@Override
	public boolean contains(@Nonnull IVec2 v) {
		boolean b1 = sign(v.x(), v.y(), point1.x, point1.y, point2.x, point2.y) < 0f;
		boolean b2 = sign(v.x(), v.y(), point2.x, point2.y, point3.x, point3.y) < 0f;
		boolean b3 = sign(v.x(), v.y(), point3.x, point3.y, point1.x, point1.y) < 0f;
		return b1 == b2 && b2 == b3;
	}

	@Override
	public boolean collides(@Nonnull Shape shape, boolean secondTry) {
		if (shape instanceof Triangle)
			return collides((Triangle)shape);
		if (shape instanceof Line)
			return collides((Line)shape);
		else
			return super.collides(shape, secondTry);
	}

	public boolean collides(@Nonnull Line line) {
		if (contains(line.point1) || contains(line.point2))
			return true;
		for (Line myLine : asPolygon().getLines()) {
			if (line.collides(myLine))
				return true;
		}
		return false;
	}

	public boolean collides(@Nonnull Triangle triangle) {
		if (contains(triangle.point1) || contains(triangle.point2) || contains(triangle.point3))
			return true;
		if (triangle.contains(point1) || triangle.contains(point2) || triangle.contains(point3))
			return true;
		for (Line line : triangle.asPolygon().getLines()) {
			if (collides(line))
				return true;
		}
		return false;
	}

	@Override
	@Nonnull public Polygon asPolygon() {
		Polygon p = new Polygon.NoHoles();
		p.addPoint(point1);
		p.addPoint(point2);
		p.addPoint(point3);
		return p;
	}

	@Override
	public void drawFilled(@Nonnull Gfx gfx, @Nonnull IVec2 v) {
		gfx.prepareShapes(ShapeRenderer.ShapeType.Filled);
		gfx.getShapeRenderer().triangle(v.x() + point1.x, v.y() + point1.y, v.x() + point2.x, v.y() + point2.y, v.x() + point3.x, v.y() + point3.y);
	}

	@Override
	public void drawOutline(@Nonnull Gfx gfx, @Nonnull IVec2 v) {
		gfx.prepareShapes(ShapeRenderer.ShapeType.Line);
		gfx.getShapeRenderer().triangle(v.x() + point1.x, v.y() + point1.y, v.x() + point2.x, v.y() + point2.y, v.x() + point3.x, v.y() + point3.y);
	}

	@Override
	@Nonnull public Triangle ease(@Nonnull Triangle other, float f) {
		return new Triangle(
				point1.ease(other.point1, f),
				point2.ease(other.point2, f),
				point3.ease(other.point3, f)
		);
	}
}

File: core/src/pl/shockah/godwit/geom/Triangle.kt
package pl.shockah.godwit.geom

import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import pl.shockah.godwit.ease.Easable
import pl.shockah.godwit.geom.polygon.ClosedPolygon
import pl.shockah.godwit.geom.polygon.Polygonable

class Triangle(
	point1: Vec2,
	point2: Vec2,
	point3: Vec2
) : Polygonable.Closed, Easable<Triangle> {
	val points: List<MutableVec2> = listOf(point1.mutableCopy(), point2.mutableCopy(), point3.mutableCopy())

	override val boundingBox: Rectangle
		get() {
			val minX = minOf(points[0].x, points[1].x, points[2].x)
			val minY = minOf(points[0].y, points[1].y, points[2].y)
			val maxX = maxOf(points[0].x, points[1].x, points[2].x)
			val maxY = maxOf(points[0].y, points[1].y, points[2].y)
			return Rectangle(ImmutableVec2(minX, minY), ImmutableVec2(maxX - minX, maxY - minY))
		}

	val lines: List<Line>
		get() = listOf(
				Line(points[0], points[1]),
				Line(points[1], points[2]),
				Line(points[2], points[3])
		)

	companion object {
		init {
			Shape.registerCollisionHandler(Triangle::class, Triangle::class) { a, b ->
				for (point in b.points) {
					if (point in a)
						return@registerCollisionHandler true
				}
				for (aLine in a.lines) {
					for (bLine in b.lines) {
						if (bLine collides aLine)
							return@registerCollisionHandler true
					}
				}
				return@registerCollisionHandler false
			}
			Shape.registerCollisionHandler(Triangle::class, Line::class) { triangle, line ->
				if (line.point1 in triangle || line.point2 in triangle)
					return@registerCollisionHandler true
				for (triangleLine in triangle.lines) {
					if (line collides triangleLine)
						return@registerCollisionHandler true
				}
				return@registerCollisionHandler false
			}
		}
	}

	override fun copy(): Triangle {
		return Triangle(points[0], points[1], points[2])
	}

	override fun equals(other: Any?): Boolean {
		return other is Triangle && points == other.points
	}

	override fun hashCode(): Int {
		return points.hashCode()
	}

	override fun translate(vector: Vec2) {
		points.forEach { it.xy += vector }
	}

	override fun mirror(horizontal: Boolean, vertical: Boolean) {
		if (horizontal)
			points.forEach { it.x = -it.x }
		if (vertical)
			points.forEach { it.y = -it.y }
	}

	override fun scale(scale: Float) {
		points.forEach { it.xy *= scale }
	}

	override operator fun contains(point: Vec2): Boolean {
		fun sign(testedPoint: Vec2, point1: Vec2, point2: Vec2): Float {
			return (testedPoint.x - point2.x) * (point1.y - point2.y) - (point1.x - point2.x) * (testedPoint.y - point2.y)
		}

		val b1 = sign(point, points[0], points[1]) < 0f
		val b2 = sign(point, points[1], points[2]) < 0f
		val b3 = sign(point, points[2], points[0]) < 0f
		return b1 == b2 && b2 == b3
	}

	override fun asClosedPolygon(): ClosedPolygon {
		return ClosedPolygon(points)
	}

	override fun ease(other: Triangle, f: Float): Triangle {
		return Triangle(
				points[0].ease(other.points[0], f),
				points[1].ease(other.points[1], f),
				points[2].ease(other.points[2], f)
		)
	}

	private fun draw(shapes: ShapeRenderer) {
		shapes.triangle(points[0].x, points[0].y, points[1].x, points[1].y, points[2].x, points[2].y)
	}

	override fun drawFilled(shapes: ShapeRenderer) {
		draw(shapes)
	}

	override fun drawOutline(shapes: ShapeRenderer) {
		draw(shapes)
	}
}